In `Multipart.from(request)`, an exception is thrown when ...

using a request with a multipart whose content is empty.

I'm not sure if this is exactly the problem, but I've created an issue because even modifying a simple existing test gives me a 500 error. 🤔 

For example:

```java
@Test
void throwsMimeParsingExceptionIfContentIsEmpty() {
    final HttpRequest request = Multipart.of(
            BodyPart.of(ContentDisposition.of("form-data", "empty"), "")
    ).toHttpRequest("/path");

    assertThatThrownBy(() -> {
        Multipart.from(request).aggregate().join();
    }).hasCauseInstanceOf(MimeParsingException.class);
}
```

I get the following exception:

```text
com.linecorp.armeria.common.multipart.MimeParsingException: java.lang.IllegalArgumentException: length 0 (expected: length > 0)
java.util.concurrent.CompletionException: com.linecorp.armeria.common.multipart.MimeParsingException: java.lang.IllegalArgumentException: length 0 (expected: length > 0)
	at java.base/java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:413)
	at java.base/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2118)
	at com.linecorp.armeria.common.util.EventLoopCheckingFuture.join(EventLoopCheckingFuture.java:87)
	at com.linecorp.armeria.common.multipart.MultipartTest.throwsMimeParsingExceptionIfContentIsEmpty(MultipartTest.java:146)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutInvocation.proceed(TimeoutInvocation.java:46)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy1/jdk.proxy1.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: com.linecorp.armeria.common.multipart.MimeParsingException: java.lang.IllegalArgumentException: length 0 (expected: length > 0)
	at app//com.linecorp.armeria.common.multipart.MimeParser.parse(MimeParser.java:290)
	at app//com.linecorp.armeria.common.multipart.MultipartDecoder.process(MultipartDecoder.java:84)
	at app//com.linecorp.armeria.internal.common.stream.DecodedStreamMessage$DecodingSubscriber.onNext(DecodedStreamMessage.java:231)
	at app//com.linecorp.armeria.common.stream.FuseableStreamMessage$FuseableSubscriber.onNext(FuseableStreamMessage.java:271)
	at app//com.linecorp.armeria.common.stream.ConcatArrayStreamMessage$ConcatArraySubscriber.onNext(ConcatArrayStreamMessage.java:177)
	at app//com.linecorp.armeria.internal.common.stream.FixedStreamMessage.onNext(FixedStreamMessage.java:214)
	at app//com.linecorp.armeria.internal.common.stream.OneElementFixedStreamMessage.request0(OneElementFixedStreamMessage.java:87)
	at app//com.linecorp.armeria.internal.common.stream.OneElementFixedStreamMessage.request(OneElementFixedStreamMessage.java:68)
	at app//com.linecorp.armeria.internal.common.stream.SubscriptionArbiter.drain(SubscriptionArbiter.java:222)
	at app//com.linecorp.armeria.internal.common.stream.SubscriptionArbiter.request0(SubscriptionArbiter.java:111)
	at app//com.linecorp.armeria.internal.common.stream.SubscriptionArbiter.request(SubscriptionArbiter.java:103)
	at app//com.linecorp.armeria.common.stream.ConcatArrayStreamMessage$ConcatArraySubscriber.request(ConcatArrayStreamMessage.java:215)
	at app//com.linecorp.armeria.common.stream.FuseableStreamMessage$FuseableSubscriber.request(FuseableStreamMessage.java:318)
	at app//com.linecorp.armeria.internal.common.stream.DecodedStreamMessage.askUpstreamForElement(DecodedStreamMessage.java:172)
	at app//com.linecorp.armeria.common.multipart.MultipartDecoder$MultipartSubscriber.requestUpstreamForBodyPartData0(MultipartDecoder.java:307)
	at app//io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at app//io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at app//io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
	at app//io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at app//io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at app//io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@17.0.2/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IllegalArgumentException: length 0 (expected: length > 0)
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:193)
	at com.linecorp.armeria.internal.common.stream.ByteBufDecoderInput.readBytes(ByteBufDecoderInput.java:139)
	at com.linecorp.armeria.common.multipart.MimeParser.readBody(MimeParser.java:352)
	at com.linecorp.armeria.common.multipart.MimeParser.parse(MimeParser.java:247)
	... 21 more
```

As I thought the problem was:

https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/common/multipart/MimeParser.java#L347
```diff 
- final ByteBuf body = in.readBytes(bodyLength);
+ final ByteBuf body = safeReadBytes(in, bodyLength);
```

Or if this is not a problem please let me know. 👍

--- a/core/src/main/java/com/linecorp/armeria/common/multipart/MimeParser.java
+++ b/core/src/main/java/com/linecorp/armeria/common/multipart/MimeParser.java
@@ -344,7 +344,7 @@ final class MimeParser {
 
             state = State.END_PART;
             done = true;
-            final ByteBuf body = in.readBytes(bodyLength);
+            final ByteBuf body = safeReadBytes(in, bodyLength);
 
             // Discard a closing boundary
             in.skipBytes(boundaryLength + 2);

--- a/core/src/test/java/com/linecorp/armeria/common/multipart/MimeParserTest.java
+++ b/core/src/test/java/com/linecorp/armeria/common/multipart/MimeParserTest.java
@@ -606,6 +606,23 @@ class MimeParserTest {
     }
 
     @Test
+    void testBodyIsEmpty() {
+        final String boundary = "boundary";
+        final byte[] chunk1 = ("--" + boundary + '\n' +
+                               "Content-Id:    \t  \t\t \n" +
+                               '\n' +
+                               "--" + boundary + "--").getBytes();
+        final List<AggregatedBodyPart> parts = parse(boundary, chunk1);
+        assertThat(parts).hasSize(1);
+
+        final AggregatedBodyPart part1 = parts.get(0);
+        assertThat(part1.headers()).hasSize(2);
+        assertThat(part1.headers().get("Content-Id")).isEqualTo("");
+        assertThat(part1.headers().contentType()).isEqualTo(MediaType.PLAIN_TEXT);
+        assertThat(part1.contentUtf8()).isEmpty();
+    }
+
+    @Test
     void testParserClosed() {
         assertThatThrownBy(() -> {
             final MimeParser parser = new MimeParser(null, null, "boundary", null);

--- a/core/src/test/java/com/linecorp/armeria/common/multipart/MultipartTest.java
+++ b/core/src/test/java/com/linecorp/armeria/common/multipart/MultipartTest.java
@@ -134,4 +134,27 @@ public class MultipartTest {
                     .hasMessageContaining("foo");
         });
     }
+
+    @Test
+    void collectContentIsEmpty() {
+        final CompletableFuture<List<Object>> collect =
+                Multipart.of(
+                        BodyPart.of(ContentDisposition.of("form-data", "name1"),
+                                    ""),
+                        BodyPart.of(ContentDisposition.of("form-data", "name2", "hello.txt"),
+                                    "")
+                ).collect(bodyPart -> {
+                    if (bodyPart.filename() != null) {
+                        final Path path = tempDir.resolve(bodyPart.name());
+                        return bodyPart.writeTo(path).thenApply(ignore -> path);
+                    }
+                    return bodyPart.aggregate().thenApply(AggregatedHttpObject::contentUtf8);
+                });
+
+        final List<Object> bodyParts = collect.join();
+        assertThat(bodyParts.get(0)).isEqualTo("");
+        final Path path = (Path) bodyParts.get(1);
+        assertThat(path).isEqualTo(tempDir.resolve("name2"));
+        assertThat(path).content().isEqualTo("");
+    }
 }

